src/Chapter3.hs

```diff
@@ -1006,9 +1156,108 @@ Implement data types and typeclasses, describing such a battle between two
+  where getNewDef :: Defence -> Defence
+        getNewDef (Defence h) = Defence (h + 5)
+
+class KnightHit a where
+  knightHit :: KnightPlayer -> a -> a
+
+instance KnightHit MonsterPlayer where
+  knightHit :: KnightPlayer -> MonsterPlayer -> MonsterPlayer
+  knightHit kp mp = mp { monsterPlayerHealth = Health(newHealthInt), monsterAlive = not isDefeated }
+    where calculateHit :: Health -> Attack -> Int
+          calculateHit (Health h) (Attack a) = h - a
+          newHealthInt = calculateHit (monsterPlayerHealth mp) (knightPlayerAttack kp)
+          isDefeated = newHealthInt <= 0
+
+instance KnightHit KnightPlayer where
+  knightHit :: KnightPlayer -> KnightPlayer -> KnightPlayer
+  knightHit kp mp = mp { knightPlayerHealth = Health(newHealthInt), knightAlive = not isDefeated }
+    where calculateHit :: Health -> Defence -> Attack -> Int
+          calculateHit (Health h) (Defence d) (Attack a) = h + d - a
+          newHealthInt = calculateHit (knightPlayerHealth mp) (knightPlayerDefence mp) (knightPlayerAttack kp)
+          isDefeated = newHealthInt <= 0
+
+class MonsterHit a where
+  monsterHit :: MonsterPlayer -> a -> a
+
+instance MonsterHit MonsterPlayer where
+  monsterHit :: MonsterPlayer -> MonsterPlayer -> MonsterPlayer
+  monsterHit kp mp = mp { monsterPlayerHealth = Health(newHealthInt), monsterAlive = not isDefeated }
+    where calculateHit :: Health -> Attack -> Int
+          calculateHit (Health h) (Attack a) = h - a
+          newHealthInt = calculateHit (monsterPlayerHealth mp) (monsterPlayerAttack kp)
+          isDefeated = newHealthInt <= 0
+
+instance MonsterHit KnightPlayer where
+  monsterHit :: MonsterPlayer -> KnightPlayer -> KnightPlayer
+  monsterHit kp mp = mp { knightPlayerHealth = Health(newHealthInt), knightAlive = not isDefeated }
+    where calculateHit :: Health -> Defence -> Attack -> Int
+          calculateHit (Health h) (Defence d) (Attack a) = h + d - a
+          newHealthInt = calculateHit (knightPlayerHealth mp) (knightPlayerDefence mp) (monsterPlayerAttack kp)
+          isDefeated = newHealthInt <= 0
+
+runAway :: MonsterPlayer -> MonsterPlayer
+runAway mp = mp { monsterAlive = False }
+
+class Alive a where
+  isAlive :: a -> Bool
+
+instance Alive MonsterPlayer where
+  isAlive :: MonsterPlayer -> Bool
+  isAlive = monsterAlive
+
+instance Alive KnightPlayer where
+  isAlive :: KnightPlayer -> Bool
+  isAlive = knightAlive
+
+-- fight :: a -> b -> b
+-- idk
+
+-- battle :: a -> b -> String
+-- battle one two 
+--   | (isAlivee one) && not (isAlivee newTwo) = "FirstWin"
+--   | not (isAlivee newOne) && (isAlivee two) = "SecondWin"
+--   | otherwise = battle newOne newTwo
+--     where
+--       newTwo = attack one two
+--       newOne = if (isAlivee newTwo)
+--         then attack two one
+--         else one
 
 {-
-You did it! Now it is time to the open pull request with your changes
+You almost did it! Now it is time to the open pull request with your changes
```

Comment:
:laughing: Don't be disappointed! I see that you did it :ok_hand: The last task is open-ended and there are multiple ways to implement it. You did a great job :clap: